src/Instrumentation/ExtRdKafka/src/ExtRdKafkaInstrumentation.php

```diff
@@ -0,0 +1,133 @@
+                $span = Span::fromContext($scope->context());
+                $span->end();
+            }
+        );
+    }
+
+    private static function addProductionHooks()
+    {
+        hook(
+            ProducerTopic::class,
+            'producev',
+            static function (
+                ProducerTopic $exchange,
+                array $params
+            ): array {
+                // Headers are the 5th argument for the producev function
+                $carrier = [];
+                TraceContextPropagator::getInstance()->inject($carrier);
+                $propagator = Globals::propagator();
+                $propagator->inject($carrier);
+
+                if (array_key_exists('traceparent', $carrier)) {
+                    $params[4]['traceparent'] = $carrier['traceparent'];
+                }
+
+                return $params;
+            }
+        );
+    }
+
+    private static function addConsumeHooks($instrumentation)
+    {
+        hook(
+            KafkaConsumer::class,
+            'consume',
+            null,
+            static function (
+                ?KafkaConsumer $exchange,
+                array $params,
+                ?Message $message,
+                ?Throwable $exception
+            ) use ($instrumentation) : void {
+                // This is to ensure that there is data. Packages periodically poll this method in order to
+                // determine if there is a message there. If there is not, we don't want to create a span.
+                if (!$message instanceof Message || $message->err !== RD_KAFKA_RESP_ERR_NO_ERROR) {
+                    return;
+                }
+
+                $builder = $instrumentation
+                    ->tracer()
+                    // @phan-suppress-next-line PhanTypeSuspiciousStringExpression - Doesn't seem to know this has to be a string
+                    ->spanBuilder('Consumer: ' . $message->topic_name)
+                    ->setSpanKind(SpanKind::KIND_CONSUMER)
+                    ->setAttribute(TraceAttributes::MESSAGING_SYSTEM, TraceAttributeValues::MESSAGING_SYSTEM_KAFKA)
+                    ->setAttribute(TraceAttributes::MESSAGING_OPERATION, TraceAttributeValues::MESSAGING_OPERATION_DELIVER)
+                    ->setAttribute(TraceAttributes::MESSAGING_KAFKA_MESSAGE_KEY, $message->key)
+                    ->setAttribute(TraceAttributes::MESSAGING_KAFKA_MESSAGE_OFFSET, $message->offset)
+                ;
+
+                $parent = Context::getCurrent();
+                $spanBuilder = $builder
+                    ->setParent($parent);
+
+                if (is_array($message->headers)) {
+                    $parent = Globals::propagator()->extract($message->headers);
+                    $spanBuilder->setParent($parent);
+                }
+
+                $span = $spanBuilder->startSpan();
+
+                $context = $span->storeInContext($parent);
+
+                Context::storage()->attach($context);
```

Comment:
Suggestion removing `setParent()` code duplication.

```suggestion
                $parent = Context::getCurrent();
                
                if (is_array($message->headers)) {
                    $parent = Globals::propagator()->extract($message->headers);
                }
                
                $spanBuilder->setParent($parent);
                $span = $spanBuilder->startSpan();

                $context = $span->storeInContext($parent);
                Context::storage()->attach($context);
```

Reply:
Just dug into this a bit. I changed it around a bit to use an `if-else`. When I did that, I got an thing from `phpstan` saying that the docblock said it would never be null.....

Did some testing (I've got this hooked into a local repo), and rdkafka does indeed give `null` as headers when there's nothing there (not empty array or anything). This is in contrast to the `php-librdkafka` stubs, which say a string.

For this reason, I've left the if statement in, and put an `ignore` on that line for `phpstan`. Not ideal, but I think that's the only way